hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/HazelcastSqlValidator.java

```diff
@@ -17,53 +17,284 @@
+            // Derive the types for offset-fetch expressions, Calcite doesn't do
+            // that automatically.
+
+            SqlSelect select = (SqlSelect) node;
+
+            SqlNode offset = select.getOffset();
+            if (offset != null) {
+                deriveType(scope, offset);
+            }
+
+            SqlNode fetch = select.getFetch();
+            if (fetch != null) {
+                deriveType(scope, fetch);
+            }
+        }
+    }
+
+    @Override
+    public SqlNode validate(SqlNode topNode) {
+        if (topNode.getKind().belongsTo(SqlKind.DDL)) {
+            return topNode;
+        }
+
+        return super.validate(topNode);
     }
 
     @Override
-    protected void addToSelectList(
-        List<SqlNode> list,
-        Set<String> aliases,
-        List<Map.Entry<String, RelDataType>> fieldList,
-        SqlNode exp,
-        SelectScope scope,
-        boolean includeSystemVars
-    ) {
+    protected void addToSelectList(List<SqlNode> list, Set<String> aliases, List<Map.Entry<String, RelDataType>> fieldList,
+                                   SqlNode exp, SelectScope scope, boolean includeSystemVars) {
         if (isHiddenColumn(exp, scope)) {
             return;
         }
 
         super.addToSelectList(list, aliases, fieldList, exp, scope, includeSystemVars);
     }
 
+    @Override
+    public RelDataType deriveType(SqlValidatorScope scope, SqlNode expression) {
+        RelDataType derived = super.deriveType(scope, expression);
+        assert derived != null;
+
+        if (derived.getSqlTypeName() == CHAR) {
+            // normalize CHAR to VARCHAR
+            derived = HazelcastTypeFactory.INSTANCE.createSqlType(VARCHAR, derived.isNullable());
+            setValidatedNodeType(expression, derived);
+        }
+
+        switch (expression.getKind()) {
+            case LITERAL:
+                return deriveLiteralType(derived, expression);
+
+            case CAST:
+                return deriveCastType(derived, scope, expression);
+
+            default:
+                return derived;
+        }
+    }
+
+    @Override
+    public void validateLiteral(SqlLiteral literal) {
+        validateLiteral(literal, getValidatedNodeType(literal));
+    }
+
+    @Override
+    public void validateCall(SqlCall call, SqlValidatorScope scope) {
+        // Enforce type derivation for all calls before validation. Calcite may
+        // skip it if a call has a fixed type, for instance AND always has
+        // BOOLEAN type, so operands may end up having no validated type.
+        deriveType(scope, call);
+        super.validateCall(call, scope);
+    }
+
+    @Override
+    protected SqlNode performUnconditionalRewrites(SqlNode node, boolean underFrom) {
+        SqlNode rewritten = super.performUnconditionalRewrites(node, underFrom);
+
+        if (rewritten != null && rewritten.isA(SqlKind.TOP_LEVEL)) {
+            // rewrite operators to Hazelcast ones starting at every top node
```

Comment:
Can we have an example of before-and-after states here, to better understand the reason of having this rewrite?

Reply:
fixed in 1c19573976bfd66fa2aa2616e22787493cacba4c